packages/next/next-server/server/config.ts

```diff
@@ -290,9 +242,69 @@ export default function loadConfig(
         )}' is not supported. Please replace the file with 'next.config.js'.`
       )
     }
+    return
   }
 
-  return defaultConfig
+  const userConfigModule = require(path)
+  const userConfig = normalizeConfig(
+    phase,
+    userConfigModule.default || userConfigModule
+  )
+  if (userConfig.target && !targets.includes(userConfig.target)) {
+    throw new Error(
+      `Specified target is invalid. Provided: "${
+        userConfig.target
+      }" should be one of ${targets.join(', ')}`
+    )
+  }
+
+  if (userConfig.amp?.canonicalBase) {
+    const { canonicalBase } = userConfig.amp || ({} as any)
+    userConfig.amp = userConfig.amp || {}
+    userConfig.amp.canonicalBase =
+      (canonicalBase.endsWith('/')
+        ? canonicalBase.slice(0, -1)
+        : canonicalBase) || ''
+  }
+
+  if (
+    userConfig.target &&
+    userConfig.target !== 'server' &&
+    ((userConfig.publicRuntimeConfig &&
+      Object.keys(userConfig.publicRuntimeConfig).length !== 0) ||
+      (userConfig.serverRuntimeConfig &&
+        Object.keys(userConfig.serverRuntimeConfig).length !== 0))
+  ) {
+    // TODO: change error message tone to "Only compatible with [fat] server mode"
+    throw new Error(
+      'Cannot use publicRuntimeConfig or serverRuntimeConfig with target=serverless https://err.sh/zeit/next.js/serverless-publicRuntimeConfig'
+    )
+  }
+
+  if (
+    userConfig.experimental?.reactMode &&
+    !reactModes.includes(userConfig.experimental.reactMode)
+  ) {
+    throw new Error(
+      `Specified React Mode is invalid. Provided: ${
+        userConfig.experimental.reactMode
+      } should be one of ${reactModes.join(', ')}`
+    )
+  }
+
+  return { configOrigin: CONFIG_FILE, ...userConfig }
+}
+
+export default function loadConfig(
+  phase: string,
+  dir: string,
+  customConfig?: object | null
+) {
+  const userConfig = loadUserConfig(phase, dir, customConfig)
+  return {
+    userConfig,
```

Comment:
I feel like exposing user configuration sets a dangerous precedent for the rest of our code base. Is there another way we can get this information? What specifically do you need this for?

If we're doing all of this to read `target`, I think we could simplify that by not setting `target: 'server'` in loadConfig and instead set the default at the callsite(s).